services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,319 @@
+            with StepProfiler(method="rows.index", step="create the row group readers"):
+                row_group_readers = [
+                    partial(parquet_file.read_row_group, i=group_id, columns=supported_columns)
+                    for parquet_file in parquet_files
+                    for group_id in range(parquet_file.metadata.num_row_groups)
+                ]
+            return row_group_offsets, row_group_readers, features, unsupported_columns
+
+    return index
+
+
+def query(offset: int, length: int, row_group_offsets: Any, row_group_readers: List[RowGroupReader]) -> pa.Table:
+    """Query the parquet files
+
+    Note that this implementation will always read one row group, to get the list of columns and always have the same
+    schema, even if the requested rows are invalid (out of range).
+
+    Args:
+        offset (int): The first row to read.
+        length (int): The number of rows to read.
+        row_group_offsets (Any): The row group offsets. See index().
+        row_group_readers (List[RowGroupReader]): The row group readers. See index().
+
+    Returns:
+        pa.Table: The requested rows.
+    """
+    if (len(row_group_offsets) == 0) or (len(row_group_readers) == 0):
+        raise ParquetResponseEmptyError("No parquet files found.")
+    last_row_in_parquet = row_group_offsets[-1] - 1
+    first_row = min(offset, last_row_in_parquet)
+    last_row = min(offset, offset + length - 1, last_row_in_parquet)
+    first_row_group_id, last_row_group_id = np.searchsorted(row_group_offsets, [first_row, last_row], side="right")
+    pa_table = pa.concat_tables([row_group_readers[i]() for i in range(first_row_group_id, last_row_group_id + 1)])
+    first_row_in_pa_table = row_group_offsets[first_row_group_id - 1] if first_row_group_id > 0 else 0
+    return pa_table.slice(offset - first_row_in_pa_table, length)
+
+
+class FeatureItem(TypedDict):
+    feature_idx: int
+    name: str
+    type: Mapping[str, Any]
+
+
+# in JSON, dicts do not carry any order, so we need to return a list
+#
+# > An object is an *unordered* collection of zero or more name/value pairs, where a name is a string and a value
+#   is a string, number, boolean, null, object, or array.
+# > An array is an *ordered* sequence of zero or more values.
+# > The terms "object" and "array" come from the conventions of JavaScript.
+# from https://stackoverflow.com/a/7214312/7351594 / https://www.rfc-editor.org/rfc/rfc7159.html
+def to_features_list(features: Features) -> List[FeatureItem]:
+    features_dict = features.to_dict()
+    return [
+        {
+            "feature_idx": idx,
+            "name": name,
+            "type": features_dict[name],
+        }
+        for idx, name in enumerate(features)
+    ]
+
+
+class RowItem(TypedDict):
+    row_idx: int
+    row: Mapping[str, Any]
+    truncated_cells: List[str]
+
+
+def to_rows_list(pa_table: pa.Table, offset: int, features: Features, unsupported_columns: List[str]) -> List[RowItem]:
```

Comment:
Should it be supported_columns?

Reply:
No: in the `index` function, we have filtered the unsupported columns (type: Audio, Image, bytes). The `query` function will not retrieve these from the parquet files. Here, we pass the list of unsupported columns to add them back to the response with empty content.

Reply:
Thanks for the clarification